enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/NamespaceEnricher.java

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright 2016 Red Hat, Inc.
+ *
+ * Red Hat licenses this file to you under the Apache License, version
+ * 2.0 (the "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package io.fabric8.maven.enricher.standard;
+
+import io.fabric8.kubernetes.api.builder.TypedVisitor;
+import io.fabric8.kubernetes.api.model.ContainerBuilder;
+import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
+import io.fabric8.maven.core.util.Configs;
+import io.fabric8.maven.core.util.kubernetes.KubernetesHelper;
+import io.fabric8.maven.enricher.api.BaseEnricher;
+import io.fabric8.maven.enricher.api.MavenEnricherContext;
+import io.fabric8.openshift.api.model.ImageChangeTriggerBuilder;
+import org.apache.commons.lang3.StringUtils;
+
+public class NamespaceEnricher extends BaseEnricher {
+    public NamespaceEnricher(MavenEnricherContext buildContext) {
+        super(buildContext, "fmp-namespace");
+    }
+
+    private enum Config implements Configs.Key {
+        namespace;
+        public String def() { return d; } protected String d;
+    }
+
+    @Override
+    public void addMissingResources(KubernetesListBuilder builder) {
+        final String namespace = getNamespace();
+
+        builder.accept(new TypedVisitor<ContainerBuilder>() {
+            @Override
+            public void visit(ContainerBuilder containerBuilder) {
+                containerBuilder.withImage(KubernetesHelper.getImageNameWithNamespace(containerBuilder.getImage(), namespace));
+            }
+        });
+
+        builder.accept(new TypedVisitor<ImageChangeTriggerBuilder>() {
+            @Override
+            public void visit(ImageChangeTriggerBuilder imageChangeTriggerBuilder) {
+                imageChangeTriggerBuilder.editOrNewFrom().withNamespace(namespace).endFrom();
+            }
+        });
+
+    }
+
+    // Get names space in the order:
+    // - plugin configuration
+    // - default name space from the kubernetes helper
+    // - "default"
+    private String getNamespace() {
+        String namespace = getConfig(Config.namespace);
+        if (StringUtils.isNotBlank(namespace)){
+            return namespace;
+        }
+
+        namespace = getContext().getConfiguration().getProperty("fabric8.namespace");
```

Comment:
With the changes from @lordofthejars the namespace also can come from the `<access>` configuration element. I'm not really sure how to deal with the namespace in generated resources. 

I mean IMO there should be no namespace included in resource descriptors which are supposed to be specific to the application and not to any target environment. Remember, the resource descriptors end up in the jar artefacts (below META-INF/) and distributed via Maven repositories. If a namespace is included here, it assumes that this namespace is available everywhere else. But this is obviously not the case.

So I really would hesitate to add the namespace directly to the resources. Instead a namespace makes sense, when _applying_ a descriptor. Or in other words, `fabric8:resource` should be namespace agnostric, and `fabric8:apply` should apply the resources in a given namespace (as it comes from the cluster access configuration). A user still can add manually a namespace to their resources if they want to (but I would highly discourage this).

We could add the enricher as optional, but would only respect the namespace given in the enricher configuration.

wdyth ?

Reply:
Ok, lets close that. I think the initial motivation coming from #1360 was that the connection to the imagestream is wrong. I thinl we could fix that differently.